Public/Src/Engine/Processes/SharedOpaqueOutputHelper.cs

```diff
@@ -10,31 +11,140 @@
+                    FileUtilities.SetFileTimestamps(expandedPath, new FileTimestamps(WellKnownTimestamps.OutputInSharedOpaqueTimestamp));
+                }
+                catch (BuildXLException ex)
+                {
+                    // On (unsafe) double writes, a race can occur, we give the other contender a second chance
+                    if (IsSharedOpaqueOutput(expandedPath))
+                    {
+                        return;
+                    }
+
+                    // Since these files should be just created outputs, this shouldn't happen and we bail out hard.
+                    throw new BuildXLException(I($"Failed to open output file '{expandedPath}' for writing."), ex);
+                }
+                finally
+                {
+                    // Restore the attributes as they were originally set
+                    if (writeAttributesDenied)
+                    {
+                        FileUtilities.SetFileAccessControl(expandedPath, FileSystemRights.WriteAttributes | FileSystemRights.WriteExtendedAttributes, allow: false);
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Checks if the given path is an output under a shared opaque by verifying whether <see cref="WellKnownTimestamps.OutputInSharedOpaqueTimestamp"/> is the creation time of the file
+            /// </summary>
+            /// <remarks>
+            /// If the given path is a directory, it is always considered part of a shared opaque
+            /// </remarks>
+            public static bool IsSharedOpaqueOutput(string expandedPath)
             {
-                // Only the creation time is used to identify a file as the output of a shared opaque
-                FileUtilities.SetFileTimestamps(expandedPath, new FileTimestamps(WellKnownTimestamps.OutputInSharedOpaqueTimestamp));
+                try
+                {
+                    var creationTime = FileUtilities.GetFileTimestamps(expandedPath).CreationTime;
+                    return creationTime == WellKnownTimestamps.OutputInSharedOpaqueTimestamp;
+                }
+                catch (BuildXLException ex)
+                {
+                    throw new BuildXLException(I($"Failed to open output file '{expandedPath}' for reading."), ex);
+                }
             }
-            catch (BuildXLException ex)
+        }
+
+        private static unsafe class Unix
+        {
+            private const string MY_XATTR_NAME = "com.microsoft.buildxl:shared_opaque_output";
+            private const long MY_XATTR_VALUE = 42;
+
+            private const int XATTR_NOFOLLOW = 1;
+
+            [DllImport("libc", EntryPoint = "setxattr")]
+            private static extern int SetXattr(
+                [MarshalAs(UnmanagedType.LPStr)] string path,
+                [MarshalAs(UnmanagedType.LPStr)] string name,
+                void *value,

```

Comment:
>void *value [](start = 16, length = 11)

can't we declare this in a more specific way? e.g. long* ?

Reply:
It doesn't have to be long*, it can be any pointer.  Here I'm simply mimicking the setxattr function signature:
```c
     int
     setxattr(const char *path, const char *name, void *value, size_t size, u_int32_t position, int options);
```
#WontFix